_gsocblogs/2025/blog_Intelligent_Logging_Pipeline_OsamaTahir.md

```diff
@@ -0,0 +1,114 @@
+</p>
+
+#### Monitor
+<p align="justify">
+The Monitor module focused on scalable storage and visualization. Loki is a distributed, horizontally scalable log aggregation system composed of several key components [2]. The Distributor receives logs from Alloy, validates them, and routes them to Ingesters while balancing the load. The Ingester temporarily stores logs in memory, compresses them, and forwards them to long-term storage in MinIO. The Querier retrieves the required logs from MinIO, forwarding them to Drain3 for prediction and Grafana for visualization. Figure 2 shows the flow of data within the customized Loki architecture.
+
+</p>
+<div align="center">
+<div style="text-align: center;">
+<img width="462" height="418" alt="custom_loki" src="https://gist.github.com/user-attachments/assets/8f29e922-ece8-425d-aca4-8f981c52e3a9" /></div>
+<p align="center"><i>Figure 2: Customized Loki</i></p>
+</div>
+
+#### Predict
+<p align="justify">
+The Predict module added intelligence to the pipeline. Drain3 parsed raw logs into structured template IDs. These template IDs are consumed by Redis in sequence from Drain3. The template ID sequences are then processed by DeepLog for sequence-based anomaly detection. Any detected anomalies are visualized by Grafana for debugging and monitoring.
+</p>
+
+### DeepLog 
+<p align="justify">
+DeepLog is an LSTM-based model that learns log patterns from normal execution [3]. It detects anomalies when log patterns deviate from the model trained from logs under normal execution. Over time, the DeepLog model adapts to new log patterns over time and constructs workflows from the underlying system log. This is because once an anomaly is detected, users can diagnose the detected anomaly and perform root cause analysis effectively. The deeplog configuration can be changed by the top K. The top K is how many of the most likely predictions will be considered "normal" by the model. If k is set to 2, you take the two events with the highest probabilities, and these will be the top k most probable next events.
+</p>
+
+<p align="justify">
+Suppose a log system has several events, each represented by a unique ID. The model takes a sequence of past events and predicts what the next event is likely to be. For example, <i>Table 1: Set of Events</i> shows unique IDs for the set of events of a user trying to upload a file. From the past event, the DeepLog learns the upcoming event and gives a probability to each unique event. It is to be noted that these sets of events are consumed by DeepLog in a random sequence.
+</p>
+
+<div align="center">
+
+| Unique ID | Event        | Probability |
+|-----------|--------------|-------------|
+| 0         | Login        | 0.7         |
+| 1         | Upload File  | 0.4         |
+| 2         | Select File  | 0.6         |
+| 3         | Logout       | 0.25        |
+| 4         | Submit File  | 0.3         |
```

Comment:
This table seems to fail to parse in the preview. It is possible that this is one of those places where github-flavored markdown is a bit different from standard. Maybe they are particular about the need for spaces:

```suggestion
| Unique ID | Event        | Probability |
| --------- | ------------ | ----------- |
| 0         | Login        | 0.7         |
| 1         | Upload File  | 0.4         |
| 2         | Select File  | 0.6         |
| 3         | Logout       | 0.25        |
| 4         | Submit File  | 0.3         |
```